File: Day15/main.py
def draw_table(data):
    if not data:
        return ""

    keys = data[0].keys()
    headers = [key.capitalize() for key in keys]

    lengths = {key: len(key) for key in keys}
    for row in data:
        for key in keys:
            lengths[key] = max(lengths[key], len(str(row[key])))

    separator = "+-" + "-+-".join("-" * lengths[key] for key in keys) + "-+"
    header_row = "| " + " | ".join(headers[i].ljust(lengths[key]) for i, key in enumerate(keys)) + "|"

    rows = []
    for row in data:
        row_content = "| " + " | ".join(str(row[key]).ljust(lengths[key]) for key in keys) + "|"
        rows.append(row_content)

    table = [separator, header_row, separator] + rows + [separator]
    return "\n".join(table)

File: Day15/main.js
function drawTable(data) {
  const keys = Object.keys(data[0])
  const headers = keys.map((key) => key.charAt(0).toUpperCase() + key.slice(1))

  const lengths = {}
  for (let key of keys) {
    lengths[key] = key.length
    for (let row of data) {
      const length = `${row[key]}`.length
      if (length > lengths[key]) {
        lengths[key] = length
      }
    }
  }

  const separator =
    '+-' + keys.map((key) => '-'.repeat(lengths[key])).join('-+-') + '-+'
  const headerRow =
    '| ' +
    keys.map((key, i) => headers[i].padEnd(lengths[key])).join(' | ') +
    ' |'

  const rows = []
  for (let row of data) {
    const rowContent =
      '| ' +
      keys.map((key) => String(row[key]).padEnd(lengths[key])).join(' | ') +
      ' |'
    rows.push(rowContent)
  }

  const table = [separator, headerRow, separator, ...rows, separator]
  return table.join('\n')
}